### DISCUSSION TITLE
Ending a trial when a condition is met (or when a process finishes)

### DISCUSSION THREAD

**Discussion by steffejr** [jspsych/jsPsych#2976] [General]
Hello,
I am working on tasks that use browser-based voice recognition. I have a simple trial that asks the user to read a sentence. This sentence is listened to and compared back to what the user reads. Voice recognition works by listening, and when the user stops speaking, it performs the sentence comparison. I want my trial to end when that process is done. Is this possible? 

I do not want a "Next" button that the user presses because they may press the button before the sentence is processed completely. And I do not want a timer since the voice recognition is cloud-based so it may be fast or slow depending on the internet connection.

Thank you

**Comment by jodeleeuw** [comment]
Hey @steffejr,

If you are writing your own plugin to handle this you can definitely do this. You just need to put the call to `jsPsych.finishTrial` in whatever function is invoked by the callback from the cloud API.

There's an example of doing this in my [pipe plugin](https://github.com/jspsych/jspsych-contrib/blob/main/packages/plugin-pipe/src/index.ts). I use `await` and wrap the plugin's run code in an `async` function. If you prefer other ways of asynchrony you can also make this work with a callback function in `fetch` or `XMLHttpRequest`.

**Reply by steffejr** [reply]
Awesome, thank you. It worked.
